### ISSUE TITLE
Kernel oops ATWILC3000 driver

### ISSUE THREAD

**Issue by ElianeBriand** [ikwzm/ZynqMP-FPGA-Ubuntu20.04-Lima-Ultra96#5]
Hello!

First, thank you a lot for your work, your Ultra96 projects are very useful!

Using the develop branch SD card files (at date 09 sept. 2021), I have successful booted of the image on the Ultra96-v2 but I get the following kernel error, and (probably as a consequence) no wifi.

First it seems like it work:

```
[    5.347788] wifi_pm : 0
[    5.350227] wifi_pm : 1
[    5.352774] wilc_sdio mmc1:0001:1: Driver Initializing success
```
Then when the network service is launched:

```
[  OK  ] Started Network Service.
[  OK  ] Found device /dev/ttyPS0.
[   10.068451] Unable to handle kernel NULL pointer dereference at virtual address 0000000000000000
[   10.077245] Mem abort info:
[   10.080034]   ESR = 0x96000005
[   10.083073]   EC = 0x25: DABT (current EL), IL = 32 bits
[   10.088366]   SET = 0, FnV = 0
[   10.091404]   EA = 0, S1PTW = 0
[   10.094528] Data abort info:
[   10.097405]   ISV = 0, ISS = 0x00000005
[   10.101230]   CM = 0, WnR = 0
[   10.104196] user pgtable: 4k pages, 39-bit VAs, pgdp=000000006b342000
[   10.110632] [0000000000000000] pgd=0000000000000000, pud=0000000000000000
[   10.117425] Internal error: Oops: 96000005 [#1] SMP
[   10.122306] Modules linked in: lima gpu_sched uio_pdrv_genirq
[   10.128070] CPU: 2 PID: 7 Comm: kworker/u8:0 Not tainted 5.4.0-xlnx-v2020.2-zynqmp-fpga-lima #1
[   10.136767] Hardware name: Avnet Ultra96-V2 Rev1 (DT)
[   10.141827] Workqueue: WILC_wq handle_get_rssi
[   10.146276] pstate: a0000085 (NzCv daIf -PAN -UAO)
[   10.151077] pc : __wake_up_common+0x58/0x170
[   10.155348] lr : __wake_up_locked+0x18/0x20
[   10.159534] sp : ffffffc01114bbe0
[   10.162842] x29: ffffffc01114bbe0 x28: 0000000000000000 
[   10.168154] x27: 0000000000000000 x26: ffffff806d408000 
[   10.173467] x25: 0000000000000001 x24: 0000000000000000 
[   10.178777] x23: 0000000000000000 x22: 0000000000000000 
[   10.184080] x21: 0000000000000003 x20: ffffff806c2670b8 
[   10.189385] x19: ffffff806c2670a8 x18: 0000000000000000 
[   10.194687] x17: 0000000000000000 x16: 0000000000000000 
[   10.199991] x15: 0000000000000000 x14: 0000000000000000 
[   10.205294] x13: 0000b56c02351a02 x12: 0000000000000001 
[   10.210598] x11: 0000000000000001 x10: 0000000000000810 
[   10.215901] x9 : ffffffc01114bd30 x8 : fefefefefefefeff 
[   10.221205] x7 : ffffffffffffffe8 x6 : 0000000000000051 
[   10.226519] x5 : 0000000000000000 x4 : 0000000000000000 
[   10.231820] x3 : 0000000000000000 x2 : 0000000000000001 
[   10.237115] x1 : 0000000000000003 x0 : 0000000000000000 
[   10.242412] Call trace:
[   10.244861]  __wake_up_common+0x58/0x170
[   10.248775]  __wake_up_locked+0x18/0x20
[   10.252595]  complete+0x6c/0xa8
[   10.255731]  wilc_wlan_cfg_commit+0x168/0x198
[   10.260092]  cfg_get+0x70/0xe0
[   10.263151]  wilc_send_config_pkt+0x140/0x1b8
[   10.267499]  handle_get_rssi+0x64/0xa0
[   10.271240]  process_one_work+0x1ec/0x370
[   10.275257]  worker_thread+0x4c/0x4f8
[   10.278914]  kthread+0x120/0x128
[   10.282149]  ret_from_fork+0x10/0x18
[   10.285722] Code: aa0503f8 f9002bfb aa0403f7 5280001b (f9400cf3) 
[   10.291815] ---[ end trace 1451a27ee044806d ]---
```

and network never completes setup.

(I have no problem using the wifi on your debian image - so the board is not damaged)

Is this a known problem?

I will try compiling a newer kernel/driver as described in doc/build/linux-kernel.md

**Comment by ikwzm** [comment]
Thank you for the issue.

Unfortunately, I have never encountered such trouble.

I'm sorry I can't be of any help.

Please let me know if you find out anything new.

**Comment by ElianeBriand** [comment]
Thank you for your response and your time. I have solved my problem.

Below is some information in case it helps someone else with the same problem:

I was using netplan (/etc/netplan/wifi.yaml) for network connection because that was the easiest way to setup wifi from the SD image without connecting the board to anything (then I would have just ssh into it). 

Unfortunately, there is probably an initialization order problem somewhere that calls on the driver too early, and produce the previously mentioned kernel oops error. This crashes the wilc driver to an unrecoverable state (once this error appears, it is not possible to use NetworkManager or anything). There are other way to connect to wifi so I did not seek the cause of that problem.

I deleted the netplan, and connected to the board using a serial line.

Using nmcli I could connect to some wifi networks without problem (`nmcli dev wifi connect SSID_of_AP password "pskkey"`)

For some reason, for other wifi it was necessary to do the connection manually. (`systemctl stop NetworkManager` then `wpa_supplicant -iwlan0 -c./wpsup&` where `wpsup` is the config file with SSID and key, then `dhclient wlan0` )
